### PR TITLE
batch.py: avoid exception when minion does not respond

### PR DESCRIPTION
**Pull request by jvanz**
### What does this PR do?

We have several issues reporting that salt is throwing exception when
the minion does not respond. This change avoid the exception adding a
default data to the minion when it fails to respond. This patch based
on the patch suggested by @roskens.

### What issues does this PR fix or reference?

Issues #46876 #48509 #50238
bsc#1135507

### Previous Behavior

salt -b 50 -L server1.domain.com,fake.domain.com test.ping 

Fail with

[ERROR   ] An un-handled exception was caught by salt's global exception handler:
KeyError: u'ret'
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 485, in salt_main
    client.run()
  File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 65, in run
    self._run_batch()
  File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 283, in _run_batch
    for res in batch.run():
  File "/usr/lib/python2.7/site-packages/salt/cli/batch.py", line 322, in run
    ret[minion] = data['ret']
KeyError: u'ret'
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 485, in salt_main
    client.run()
  File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 65, in run
    self._run_batch()
  File "/usr/lib/python2.7/site-packages/salt/cli/salt.py", line 283, in _run_batch
    for res in batch.run():
  File "/usr/lib/python2.7/site-packages/salt/cli/batch.py", line 322, in run
    ret[minion] = data['ret']
KeyError: u'ret'

### New Behavior

\# salt --out=raw  -L germ109.suse.cz,fake.domain.cz test.ping 
{u'germ109.suse.cz': True}
{u'fake.domain.cz': u'Minion did not return. [Failed]'}

### Tests written?

No

### Commits signed with GPG?

No

### CODE REVIEW

**Comment by jvanz** [comment]
I don't know why the the jenkins/pr/py3-centos-7 is failing. As far as I can see, the job is failing because a problem in the python packages installation. How can I fix that?

**Review by meaksh — APPROVED** [review]

File: salt/cli/batch.py [hunk 1]
```diff
@@ -312,6 +312,11 @@ def run(self):
                     if self.opts.get('failhard') and data['ret']['retcode'] > 0:
                         failhard = True
 
+                # avoid an exception if the minion does not respond.
+                if data.get("failed") is True:
+                    log.debug('Minion %s failed to respond: data=%s', minion, data)
+                    data = {'ret': 'Minion did not return. [Failed]', 'retcode': 1}
```
Comment (DmitryKuzmenko) [inline salt/cli/batch.py]: I'm ok except the hardcoded retcode value. Could you please use a constant instead?
```
'retcode': salt.defaults.exitcodes.EX_GENERIC
```

**Review by DmitryKuzmenko — CHANGES_REQUESTED** [review]

File: salt/cli/batch.py (hunk 1 above)
Comment (jvanz) [reply salt/cli/batch.py]: Indeed. You're right. I forgot to search for a constant.

Comment (jvanz) [reply salt/cli/batch.py]: I've pushed a new commit fixing this

**Review by DmitryKuzmenko — APPROVED** [review]

**Comment by dwoz** [comment]
re-run full all

File: salt/cli/batch.py [hunk 2]
```diff
@@ -312,6 +312,11 @@ def run(self):
                     if self.opts.get('failhard') and data['ret']['retcode'] > 0:
                         failhard = True
 
+                # avoid an exception if the minion does not respond.
+                if data.get("failed") is True:
+                    log.debug('Minion %s failed to respond: data=%s', minion, data)
+                    data = {'ret': 'Minion did not return. [Failed]', 'retcode': salt.defaults.exitcodes.EX_GENERIC}
```
Comment (dincamihai) [inline salt/cli/batch.py]: I've just noticed that here: https://github.com/saltstack/salt/blob/e81e495b2713336f9908e5f4cc15ba8ebcb67306/salt/cli/batch.py#L311 the structure of 'ret' is different. Shouldn't this be unified?

Comment (dincamihai) [reply salt/cli/batch.py]: I'm also not sure if completely replacing `data` with a new dictionary is the right thing to do here.

**Review by dincamihai — CHANGES_REQUESTED** [review]

File: salt/cli/batch.py (hunk 2 above)
Comment (DmitryKuzmenko) [reply salt/cli/batch.py]: @dincamihai 

> the structure of 'ret' is different. Shouldn't this be unified?

The snipped you've mentioned is applied in the only case the executed module returns a `dict` value. In  other cases the return value is cusually `bool`. I the error case this PR handles the return value is absent and it makes it a string. So `retcode` is put in the correct place here.

> I'm also not sure if completely replacing data with a new dictionary is the right thing to do here.

This change is making batch mode work similar to non-batch. In the described case non-batch return is implemented here: https://github.com/saltstack/salt/blob/develop/salt/client/__init__.py#L1580 it just returns a dict containing `out`, `ret` and `retcode` values. Probably it'd be more correct to add `'out': 'no_return'` here too. But overall this PR looks good.

**Review by DmitryKuzmenko — APPROVED** [review]

File: salt/cli/batch.py (hunk 2 above)
Comment (dincamihai) [reply salt/cli/batch.py]: > The snipped you've mentioned is applied in the only case the executed module returns a dict value. In other cases the return value is cusually bool. I the error case this PR handles the return value is absent and it makes it a string. So retcode is put in the correct place here.

You are right, but shouldn't we align that exception and have an unified return structure? The structure should not change depending on the executed module.

**Review by jvanz — COMMENTED** [review]
Why do we have the second pull request, #146? Sorry, I don't get it.

File: salt/cli/batch.py (hunk 2 above)
Comment (jvanz) [reply salt/cli/batch.py]: Guys, sorry but I'm still confused. Should I change the return dict? It is still no clear to me what was the decision. =/

**Comment by bryceml** [comment]
Having an unknown repository as a source breaks jenkins.  If you would like this PR to make it in, please re-submit and link to the original.

**Pull request closed without merging**